### PR TITLE
feat: Add Bulk Subscribe support

### PR DESCRIPTION
**Pull request by DeepanshuA**
Signed-off-by: Deepanshu Agarwal <deepanshu.agarwal1984@gmail.com>

# Description

This PR intends to add Bulk Subscribe Support to js-sdk.
Currently, it is using dapr/dapr master as still awaiting Dapr v1.10 release, which when used will contain bulk subscribe support in runtime.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close a part of: #360 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Extended the documentation

### CODE REVIEW

**Comment by XavierGeerinck** [comment]
Discussed during weekly: add tests + docs so we can review

**Comment by DeepanshuA** [comment]
> Discussed during weekly: add tests + docs so we can review

Done @XavierGeerinck @shubham1172 - This PR is ready for review now.

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 1]
```diff
@@ -280,6 +280,43 @@ async function start() {
 }
 ```
 
+#### Bulk Subscribe to messages
+
+Dapr [supports bulk subscribing](https://v1-10.docs.dapr.io/developing-applications/building-blocks/pubsub/pubsub-bulk/) a topic.
```
Comment (XavierGeerinck) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:285]: Bulk Subscription is supported and supports two flavors in its API:

- `bulkSubscribeWithDefaultConfig`: Bulk Subscribe to a topic, utilising default values for `maxMessagesCount` and `maxAwaitDurationMs` 
- `bulkSubscribeWithConfig`: Bulk Subscribe to a topic and support setting the `maxMessagesCount` and `maxAwaitDurationMs`

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 2]
```diff
@@ -280,6 +280,43 @@ async function start() {
...
+
```
Comment (XavierGeerinck) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:286]: In general it feels redundant to have both these methods? Can't we make these parameters optional?

Also, can we name them `subscribeBulk` instead? making the API become: `server.pubsub.subscribeBulk` which follows the SDK a bit more

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 3]
```diff
@@ -280,6 +280,43 @@ async function start() {
...
+  await client.pubsub.bulkSubscribeWithDefaultConfig(pubSubName, topic, (data) =>
+    console.log("Subscriber received: " + JSON.stringify(data)),
+  );
+
+  // Publish multiple messages to a topic with specific maxMessagesCount and maxAwaitDurationMs.
+  await client.pubsub.bulkSubscribeWithConfig(
+    pubSubName,
+    topic,
+    (data) => console.log("Subscriber received: " + JSON.stringify(data)),
+    100,
```
Comment (XavierGeerinck) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:314]: Ref latest API spec changes, maybe make it in a parameters object with defaults?

File: examples/pubsub/src/index.ts [hunk 4]
```diff
@@ -29,6 +29,29 @@ async function start() {
...
+
+  // Publish multiple messages to a topic with specific maxMessagesCount and maxAwaitDurationMs.
+  await client.pubsub.bulkSubscribeWithConfig(
+    "my-pubsub-component",
+    "my-topic",
+    async (data: Record<string, any>) => {
+      // The library parses JSON when possible.
+      console.log(`[Dapr-JS][Example] Received on subscription: ${JSON.stringify(data)}`);
+    },
+    100,
```
Comment (XavierGeerinck) [inline examples/pubsub/src/index.ts:51]: Ref latest API spec changes, maybe make it in a parameters object with defaults?

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 5]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+      let data: any;
+      if (event.hasBytes()) {
+        data = Buffer.from(event.getBytes()).toString();
+      } else if (event.hasCloudEvent()) {
+        const cloudEvent = event.getCloudEvent();
+        if (cloudEvent instanceof TopicEventCERequest) {
+          data = Buffer.from(cloudEvent.getData()).toString();
+        }
+      }
+      let dataParsed: any;
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:495]: Should be merged with PR #453 to abstract the extraction logic

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 6]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+      );
+      return;
+    }
+
+    const resArr: TopicEventBulkResponseEntry[] = [];
+    const entries = req.getEntriesList();
+    for (const ind in entries) {
+      const event = entries[ind];
+      let data: any;
+      if (event.hasBytes()) {
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:487]: Should be merged with PR #453 to utilize Deserialization correctly

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 7]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+    ) {
+      this.logger.warn(
+        `The topic '${topic}' is not being subscribed to on PubSub '${pubsubName}' for route '${route}'.`,
+      );
+      return;
+    }
+
+    const resArr: TopicEventBulkResponseEntry[] = [];
+    const entries = req.getEntriesList();
+    for (const ind in entries) {
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:484]: Newline

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 8]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+          data = Buffer.from(cloudEvent.getData()).toString();
+        }
+      }
+      let dataParsed: any;
+      try {
+        dataParsed = JSON.parse(data);
+      } catch (e) {
+        dataParsed = data;
+      }
+      const res = new TopicEventBulkResponseEntry();
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:501]: newline

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 9]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+
+      try {
+        const eventHandlers = this.pubSubSubscriptions[pubsubName][topic].routes[route].eventHandlers;
+        await Promise.all(eventHandlers.map((cb) => cb(dataParsed, headers)));
+        res.setStatus(TopicEventResponse.TopicEventResponseStatus.SUCCESS);
+      } catch (e) {
+        this.logger.error(`Error handling topic event: ${e}`);
+        res.setStatus(TopicEventResponse.TopicEventResponseStatus.DROP);
+      }
+      res.setEntryId(event.getEntryId());
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:520]: newline

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 10]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+
+      for (const [key, value] of Object.entries(call.metadata.toHttp2Headers())) {
+        if (value) {
+          headers[key] = value.toString();
+        }
+      }
+
+      try {
+        const eventHandlers = this.pubSubSubscriptions[pubsubName][topic].routes[route].eventHandlers;
+        await Promise.all(eventHandlers.map((cb) => cb(dataParsed, headers)));
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:514]: To be merged with PR #453 for event handlers abstraction

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 11]
```diff
@@ -449,6 +457,75 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+        this.logger.error(`Error handling topic event: ${e}`);
+        res.setStatus(TopicEventResponse.TopicEventResponseStatus.DROP);
+      }
+      res.setEntryId(event.getEntryId());
+      resArr.push(res);
+    }
+
+    const totalRes = new TopicEventBulkResponse();
+    totalRes.setStatusesList(resArr);
+    return callback(null, totalRes);
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:526]: newline

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 12]
```diff
@@ -471,6 +548,18 @@ export default class GRPCServerImpl implements IAppCallbackServer {
           topicSubscription.setDeadLetterTopic(daprConfig.deadLetterTopic);
         }
 
+        if (daprConfig?.bulkSubscribe) {
+          const bulkSubscribe = new BulkSubscribeConfig();
+          bulkSubscribe.setEnabled(daprConfig.bulkSubscribe.enabled);
+          if (daprConfig?.bulkSubscribe?.maxMessagesCount) {
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:554]: newline

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts [hunk 13]
```diff
@@ -471,6 +548,18 @@ export default class GRPCServerImpl implements IAppCallbackServer {
...
+            bulkSubscribe.setMaxMessagesCount(daprConfig.bulkSubscribe.maxMessagesCount);
+          }
+          if (daprConfig?.bulkSubscribe?.maxAwaitDurationMs) {
+            bulkSubscribe.setMaxAwaitDurationMs(daprConfig.bulkSubscribe.maxAwaitDurationMs);
+          }
+          topicSubscription.setBulkSubscribe(bulkSubscribe);
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/GRPCServerImpl.ts:560]: newline

File: src/implementation/Server/GRPCServer/pubsub.ts [hunk 14]
```diff
@@ -97,6 +98,48 @@ export default class DaprPubSub implements IServerPubSub {
...
+    pubsubName: string,
+    topic: string,
+    cb: TypeDaprPubSubCallback,
+    route: string | DaprPubSubRouteType = "",
+    metadata?: KeyValueType,
+  ): Promise<void> {
+    const bulkSubscribe: BulkSubscribeConfig = {
+      enabled: true,
+    };
+    this.server
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/pubsub.ts:111]: newline

File: src/implementation/Server/GRPCServer/pubsub.ts [hunk 15]
```diff
@@ -97,6 +98,48 @@ export default class DaprPubSub implements IServerPubSub {
...
+    maxAwaitDurationMs: number,
+    route: string | DaprPubSubRouteType = "",
+    metadata?: KeyValueType,
+  ): Promise<void> {
+    const bulkSubscribe: BulkSubscribeConfig = {
+      enabled: true,
+      maxMessagesCount: maxMessagesCount,
+      maxAwaitDurationMs: maxAwaitDurationMs,
+    };
+    this.server
```
Comment (XavierGeerinck) [inline src/implementation/Server/GRPCServer/pubsub.ts:134]: newline

File: src/implementation/Server/HTTPServer/HTTPServerImpl.ts [hunk 16]
```diff
@@ -120,6 +123,11 @@ export default class HTTPServerImpl {
 
       // Add a server POST handler
       this.server.post(`/${routeObj.path}`, async (req, res) => {
+        const bulkSubEnabled = this.pubSubSubscriptions[pubsubName][topic].dapr.bulkSubscribe?.enabled;
+        if (bulkSubEnabled) {
+          const result = await this.processBulkSubscribeMessage(routeObj, req);
+          return res.send(result);
+        }
         const headers = req.headers;
```
Comment (XavierGeerinck) [inline src/implementation/Server/HTTPServer/HTTPServerImpl.ts:131]: newline

File: test/e2e/common/server.test.ts [hunk 17]
```diff
@@ -109,6 +117,112 @@ describe("common/server", () => {
...
+    runIt(
+      "should be able to publish multiple ce messages but receive rawPayload events using bulk subscribe",
+      async (server: DaprServer, protocol: string) => {
+        const res1 = await server.client.pubsub.publish(pubSubName, getTopic(bulkSubscribeCE_RPTopic, protocol), {
+          message: "Hello, world1111!",
+        });
+        const res2 = await server.client.pubsub.publish(pubSubName, getTopic(bulkSubscribeCE_RPTopic, protocol), {
+          message: "Hello, world2222!",
+        });
+        expect(res1.error).toBeUndefined();
```
Comment (XavierGeerinck) [inline test/e2e/common/server.test.ts:181]: newline

File: test/e2e/common/server.test.ts [hunk 18]
```diff
@@ -417,6 +531,58 @@ describe("common/server", () => {
...
+    );
+
+    await grpcServer.pubsub.bulkSubscribeWithDefaultConfig(
+      pubSubName,
+      getTopic(bulkSubscribeCE_RPTopic, protocolGrpc),
+      mockBulkSubscribeCE_RPHandler,
+      undefined,
+      { rawPayload: true },
+    );
+    await httpServer.pubsub.bulkSubscribeWithDefaultConfig(
```
Comment (XavierGeerinck) [inline test/e2e/common/server.test.ts:567]: Newline

File: test/e2e/common/server.test.ts [hunk 19]
```diff
@@ -417,6 +531,58 @@ describe("common/server", () => {
...
+      undefined,
+      { rawPayload: true },
+    );
+
+    await grpcServer.pubsub.bulkSubscribeWithDefaultConfig(
+      pubSubName,
+      getTopic(bulkSubscribeRP_CETopic, protocolGrpc),
+      mockBulkSubscribeRP_CEHandler,
+    );
+    await httpServer.pubsub.bulkSubscribeWithDefaultConfig(
```
Comment (XavierGeerinck) [inline test/e2e/common/server.test.ts:580]: Newline

**Review by XavierGeerinck — CHANGES_REQUESTED** [review]
Major things:
- Check around API spec renaming towards `subscribeBulk` and adding a default params object
- Ensure correct newline usage

File: .github/workflows/test-e2e.yml [hunk 20]
```diff
@@ -39,7 +39,7 @@ jobs:
       DAPR_RUNTIME_VER: 1.9.5
       DAPR_INSTALL_URL: https://raw.githubusercontent.com/dapr/cli/master/install/install.sh
       DAPR_CLI_REF: ""
-      DAPR_REF: ""
+      DAPR_REF: "2accc6c5ff16b29d8e49fe3e02487b0e959aa0f2"
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply .github/workflows/test-e2e.yml:42]: No, this is required as else the tests will fail with Dapr 1.9.5. This will be removed once we start using Dapr 1.10.

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 2 above)
Comment (DeepanshuA) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:286]: Done

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts (hunk 7 above)
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/GRPCServerImpl.ts:485]: Done

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts (hunk 8 above)
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/GRPCServerImpl.ts:503]: Done

File: src/implementation/Server/GRPCServer/pubsub.ts [hunk 21]
```diff
@@ -97,6 +98,48 @@ export default class DaprPubSub implements IServerPubSub {
     this.server.getServerImpl().registerPubSubSubscriptionEventHandler(pubsubName, topic, route, cb);
   }
 
+  async bulkSubscribeWithDefaultConfig(
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/pubsub.ts]: Done

File: src/implementation/Server/GRPCServer/pubsub.ts [hunk 22]
```diff
@@ -97,6 +98,48 @@ export default class DaprPubSub implements IServerPubSub {
...
+    this.server
+      .getServerImpl()
+      .registerPubsubSubscription(pubsubName, topic, { route, metadata, bulkSubscribe: bulkSubscribe });
+
+    // Add the callback to the event handlers manually
+    // @todo: we will deprecate this way of working? and require subscribeToRoute?
+    this.subscribeToRoute(pubsubName, topic, route, cb);
+  }
+
+  async bulkSubscribeWithConfig(
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/pubsub.ts]: Done

File: src/implementation/Server/HTTPServer/HTTPServerImpl.ts [hunk 23]
```diff
@@ -120,6 +123,11 @@ export default class HTTPServerImpl {
 
       // Add a server POST handler
       this.server.post(`/${routeObj.path}`, async (req, res) => {
+        const bulkSubEnabled = this.pubSubSubscriptions[pubsubName][topic].dapr.bulkSubscribe?.enabled;
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/implementation/Server/HTTPServer/HTTPServerImpl.ts:126]: bulkSubscribe has been used here with Optional Chaining only.

File: src/implementation/Server/HTTPServer/pubsub.ts [hunk 24]
```diff
@@ -97,6 +98,48 @@ export default class HTTPServerPubSub implements IServerPubSub {
...
+    this.server
+      .getServerImpl()
+      .registerPubsubSubscription(pubsubName, topic, { route, metadata, bulkSubscribe: bulkSubscribe });
+
+    // Add the callback to the event handlers manually
+    // @todo: we will deprecate this way of working? and require subscribeToRoute?
+    this.subscribeToRoute(pubsubName, topic, route, cb);
+  }
+
+  async bulkSubscribeWithConfig(
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/implementation/Server/HTTPServer/pubsub.ts]: Done

File: src/interfaces/Server/IServerPubSub.ts [hunk 25]
```diff
@@ -60,6 +60,48 @@ export default interface IServerPubSub {
...
+   * @param pubSubName name of the pubsub
+   * @param topic name of the topic
+   * @param cb callback function to handle messages
+   * @param maxMessagesCount max messages count
+   * @param maxAwaitDurationMs max await duration in milliseconds
+   * @param route The HTTP route override to register for the event subscription.
+   * Default value is `/route-${pubsubName}-${topic}`. Ignored if gRPC is used.
+   * @param metadata metadata for the subscription
+   */
+  bulkSubscribeWithConfig(
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/interfaces/Server/IServerPubSub.ts]: Done

File: examples/pubsub/src/index.ts (hunk 4 above)
Comment (DeepanshuA) [reply examples/pubsub/src/index.ts]: Done

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 1 above)
Comment (DeepanshuA) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md]: Only one flavor supported now, as per further review comments.

File: src/interfaces/Server/IServerPubSub.ts [hunk 26]
```diff
@@ -60,6 +60,48 @@ export default interface IServerPubSub {
...
+   * are not provided and default values for related component will be used.
+   *
+   * @param pubSubName name of the pubsub
+   * @param topic name of the topic
+   * @param cb callback function to handle messages
+   * @param route The HTTP route override to register for the event subscription.
+   * Default value is `/route-${pubsubName}-${topic}`. Ignored if gRPC is used.
+   * @param metadata metadata for the subscription
+   */
+  bulkSubscribeWithDefaultConfig(
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/interfaces/Server/IServerPubSub.ts]: Done

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts (hunk 5 above)
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/GRPCServerImpl.ts]: Done

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts (hunk 6 above)
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/GRPCServerImpl.ts:493]: Done

File: src/implementation/Server/GRPCServer/GRPCServerImpl.ts (hunk 10 above)
Comment (DeepanshuA) [reply src/implementation/Server/GRPCServer/GRPCServerImpl.ts]: Done

File: src/implementation/Server/HTTPServer/HTTPServerImpl.ts [hunk 27]
```diff
@@ -146,6 +154,57 @@ export default class HTTPServerImpl {
     );
   }
 
+  async processBulkSubscribeMessage(
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply src/implementation/Server/HTTPServer/HTTPServerImpl.ts:150]: Yeah, it will be still required, as bulk subscribe has git its own processing logic. But, inside it, I was able to re-use process of call back method and retry logic.

**Comment by DeepanshuA** [comment]
> Major things:
> 
> * Check around API spec renaming towards `subscribeBulk` and adding a default params object
> * Ensure correct newline usage

Incorporated review comments. Please review again.

File: test/e2e/common/server.test.ts [hunk 28]
```diff
@@ -109,6 +117,112 @@ describe("common/server", () => {
...
+        expect(mockBulkSubscribeCEHandler.mock.calls.length).toBe(2);
+        expect(mockBulkSubscribeCEHandler.mock.calls[0][0]).toEqual({ message: "Hello, world1111!" });
+        expect(mockBulkSubscribeCEHandler.mock.calls[1][0]).toEqual({ message: "Hello, world2222!" });
+      },
+    );
+
+    runIt(
+      "should be able to publish multiple ce messages but receive rawPayload events using bulk subscribe",
+      async (server: DaprServer, protocol: string) => {
+        const res1 = await server.client.pubsub.publish(pubSubName, getTopic(bulkSubscribeCE_RPTopic, protocol), {
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply test/e2e/common/server.test.ts]: Done

File: components/pubsub.yaml [hunk 29]
```diff
@@ -0,0 +1,11 @@
...
+kind: Component
+metadata:
+  name: pubsub
+spec:
+  type: pubsub.redis
+  metadata:
+  - name: redisHost
+    value: localhost:6379
+  - name: redisPassword
+    value: ""
```
Comment (shubham1172) [inline components/pubsub.yaml:11]: What are these components used for? If these are for tests, they should be moved inside `test/components`.

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 30]
```diff
@@ -351,6 +351,46 @@ async function start() {
...
+const pubSubName = "orderPubSub";
+const topic = "topicbulk";
+
+const DAPR_HOST = process.env.DAPR_HOST || "127.0.0.1";
+const DAPR_HTTP_PORT = process.env.DAPR_HTTP_PORT || "3502";
+const SERVER_HOST = process.env.SERVER_HOST || "127.0.0.1";
+const SERVER_PORT = process.env.APP_PORT || 5001;
+
+async function start() {
+  const server = new DaprServer(SERVER_HOST, SERVER_PORT, DAPR_HOST, DAPR_HTTP_PORT);
```
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:374]: TODO: Update this once #462 is merged.

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 31]
```diff
@@ -351,6 +351,46 @@ async function start() {
...
+  // Publish multiple messages to a topic with default config.
+  await client.pubsub.subscribeBulk(pubSubName, topic, (data) =>
+    console.log("Subscriber received: " + JSON.stringify(data)),
+  );
+
+  // Publish multiple messages to a topic with specific maxMessagesCount and maxAwaitDurationMs.
+  await client.pubsub.subscribeBulk(
+    pubSubName,
+    topic,
+    (data) => console.log("Subscriber received: " + JSON.stringify(data)),
```
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:385]: Can we show here what is in the data? Maybe show how to iterate over to get individual messages and prepare a response back (and what is the default).

**Review by shubham1172 — CHANGES_REQUESTED** [review]

File: test/e2e/common/server.test.ts [hunk 32]
```diff
@@ -196,6 +204,128 @@ describe("common/server", () => {
...
+          message: "Hello, world1111!",
+        });
+        expect(getDataFromCEObject(mockBulkSubscribeCloudEventToRawPayloadHandler.mock.calls[1][0])).toEqual({
+          message: "Hello, world2222!",
+        });
+      },
+    );
+
+    runIt(
+      "should be able to publish multiple rawPayload messages but receive ce events using bulk subscribe",
```
[the remark this answers is not in the document]
Comment (DeepanshuA) [reply test/e2e/common/server.test.ts]: Done

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 30 above)
Comment (DeepanshuA) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:376]: 462 will take care of it, I plan to get this PR of Bulk Sub merged first.

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md (hunk 31 above)
Comment (DeepanshuA) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md:387]: Updated comments in the lines above to specify details of callback.

File: components/pubsub.yaml (hunk 29 above)
Comment (DeepanshuA) [reply components/pubsub.yaml]: My bad, I included them by mistake - Removed now.

File: test/e2e/common/server.test.ts [hunk 33]
```diff
@@ -196,6 +204,128 @@ describe("common/server", () => {
...
     });
 
+    runIt(
+      "should be able to publish multiple rawPayload messages and receive events using bulk subscribe",
+      async (server: DaprServer, protocol: string) => {
+        const res1 = await server.client.pubsub.publish(
+          pubSubName,
+          getTopic(bulkSubscribeTopic, protocol),
+          {
+            message: "Hello, world1111!",
```
Comment (shubham1172) [inline test/e2e/common/server.test.ts:214]: This is just a nit
```suggestion
            message: "Message 1",
```

File: src/types/pubsub/BulkSubscribeOptions.type.ts [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+ * BulkSubscribeOptions enlists the options for bulk subscribe
+ **/
+export type BulkSubscribeOptions = {
+  // The route creation for a single route or DaprPubSubRouteType
+  route?: string | DaprPubSubRouteType;
+
+  // Metadata for the subscription
+  metadata?: KeyValueType;
+
+  //Max number of messages to be sent in a single bulk request
```
Comment (shubham1172) [inline src/types/pubsub/BulkSubscribeOptions.type.ts:27]:
```suggestion
  // Max number of messages to be sent in a single bulk request
```
Nit: Other places too.

File: test/e2e/common/server.test.ts (hunk 33 above)
Comment (DeepanshuA) [reply test/e2e/common/server.test.ts]: Done

File: daprdocs/content/en/js-sdk-docs/js-server/_index.md [hunk 35]
```diff
@@ -351,6 +351,48 @@ async function start() {
 }
 ```
 
+#### Bulk Subscribe to messages
+
+Bulk Subscription is supported and is available through following API:
+
+- Bulk subscription through the `subscribeBulk` method: `maxMessagesCount` and `maxAwaitDurationMs` are optional; and if not provided, default values for related components will be used.
+
+App need not iterate through multiple entries inside a received bulk subscribe event. Callback provided by App is applicable per Entry. And, similarly, while writing callback, App can decide to send back SUCCESS/RETRY/DROP, as defined above.
```
Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:360]: While listening for messages, the application receives messages from Dapr in bulk. However, like regular subscribe, the callback function receives a single message at a time, and the user can choose to return a `DaprPubSubStatusEnum` value to acknowledge successfully, retry, or drop the message. The default behavior is to return a success response.

Comment (shubham1172) [inline daprdocs/content/en/js-sdk-docs/js-server/_index.md:360]:
```suggestion
While listening for messages, the application receives messages from Dapr in bulk. However, like regular subscribe, the callback function receives a single message at a time, and the user can choose to return a `DaprPubSubStatusEnum` value to acknowledge successfully, retry, or drop the message. The default behavior is to return a success response.
```

Comment (DeepanshuA) [reply daprdocs/content/en/js-sdk-docs/js-server/_index.md]: Done and added return type in a callback with specifying that as per different processing, can return RETRY/DROP/SUCCESS.

**Comment by DeepanshuA** [comment]
Removal for Review is NOT intentional, when I request re-review from someone, other requests get cancelled somehow.

File: src/types/pubsub/BulkSubscribeConfig.type.ts [hunk 36]
```diff
@@ -0,0 +1,28 @@
...
+// https://github.com/dapr/dapr/blob/master/pkg/apis/subscriptions/v2alpha1/types.go#L53
+
+/**
+ * BulkSubscribeConfig defines the configuration for a bulk subscription
+ **/
+export type BulkSubscribeConfig = {
+  // Flag to enable/disable bulk subscribe
+  enabled: boolean;
+
+  //Max number of messages to be sent in a single bulk request
```
Comment (shubham1172) [inline src/types/pubsub/BulkSubscribeConfig.type.ts:23]: Same 
```suggestion
  // Max number of messages to be sent in a single bulk request
```

File: src/types/pubsub/BulkSubscribeConfig.type.ts [hunk 37]
```diff
@@ -0,0 +1,28 @@
...
+  maxMessagesCount?: number;
+
+  //Max duration to wait for messages to be sent in a single bulk request
```
Comment (shubham1172) [inline src/types/pubsub/BulkSubscribeConfig.type.ts:26]:
```suggestion
  // Max duration to wait for messages to be sent in a single bulk request
```

File: src/proto/dapr/proto/runtime/v1/dapr.proto
Comment (shubham1172) [inline src/proto/dapr/proto/runtime/v1/dapr.proto:1]: This is an unrelated and **unreleased** change, correct?

Comment (shubham1172) [reply src/proto/dapr/proto/runtime/v1/dapr.proto:1]: https://github.com/dapr/dapr/blob/2651bef6736f6d48d764e106a3a7ac597d889f12/dapr/proto/runtime/v1/dapr.proto#L383 v1.10.3-rc2 does not contain it.

Comment (DeepanshuA) [reply src/proto/dapr/proto/runtime/v1/dapr.proto:1]: Nice catch. Yeah, this proto should not have been here. Removed it.

**Review by XavierGeerinck — APPROVED** [review]
